asynclocalstorage.md

```diff
@@ -0,0 +1,371 @@
+by Web Platform APIs (e.g. timers, background tasks, unhandled rejections),
+some defined by the runtime, and others defined by the application. The
+common characterist of all async resources is that they are expected to
+capture the current async context frame and propagate it at various
+specific times.
+
+### Promises
+
+A promise itself is *not* an async resource. A promise continuation task,
+however, *is*. In other words, when we use `then()` to attach a continuation
+task to a promise, the current async context is captured and associated with
+that continuation task.
+
+(Note that this is a variance from Node.js' current implementation which treats
+the promise *itself* as the async resource, capturing the context at the moment
+the promise is created rather than when the continuation task is created.)
+
+#### Unhandled rejections and the deferred promise pattern
+
+The `unhandledrejection` and `rejectionhandled` events should, by default,
+propagate the context that was current when the promise was rejected and *not*
+when the promise was created.
+
+(Note that these are the only standard `EventTarget` events that are expected
+to propagate async context currently. All other events should be invoked in
+the same frame that the `dispatchEvent()` method was called in.)
+
+The deferred promise pattern is an approach to using promises that separates
+the `resolve()` and `reject()` functions used to settle the promise from the
+actual promise itself.
+
+```js
+function deferredPromise() {
+  let resolve, reject;
+  const promise = new Promise((a,b) => {
+    resolve = a;
+    reject = b;
+  });
+  return { promise, resolve, reject }
+}
+```
+
+Care must be taken when using this pattern to properly propagate the async
+context to the Web API `unhandledrejection` and `rejectionhandled` events.
+
+Specifically, given the following example, the `unhandledrejection` event
+should, by default, get the value `321` from the `AsyncLocalStorage`
+instance, rather than `123`:
+
+```js
+const als = new AsyncLocalStorage();
+
+addEventListener('unhandledrejection', (event) => {
+  console.log(als.getStore()); //321
+});
+
+const { promise, reject } = als.run(123, () => deferredPromise());
+als.run(321, () => reject('boom'));
+```
+
+If someone really wants the `unhandledrejection` event to receive the
+context at the time the promise was created, then the `reject()` should be
+wrapped using `AsyncResource.bind()`, for instance:
+
+```js
+function deferredPromiseCurrentContext() {
+  let resolve, reject;
+  const promise = new Promise((a,b) => {
+    resolve = AsyncResource.bind(a);
+    reject = AsyncResource.bind(b);
+  });
+  return { promise, resolve, reject }
+}
+```
+
+This would ensure that whenever `reject()` is called, it enters the
+context that was current when the promise was created.
+
+#### Thenables
+
+Thenables (promise-like objects that expose a `then()` method, should be handled
```

Comment:
Is this to say that V8's behavior is buggy? Or, that libraries and runtimes that expose thenables have a responsibility to use AsyncResource appropriately? Maybe be slightly clearer here.

Reply:
Probably a matter of debate, but V8 has historical not treated thenables quite the same. I had the same problem with promise hooks and had to make some changes to V8 to get it to emit promise hook events for thenables too.